Add slotFill for variation menus

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds slotFills for the `Quick update` menus to enhance extensibility. It also refactors them to remove some code duplication we had.

### Extensibility

**_Quick update_ actions**
```
<VariationQuickUpdateMenuItem
    group= { 'pricing' }
    supportsMultipleSelection={ true }
    order={ 20 }
>
    // Dropdown, MenuGroup, MenuItem
</VariationQuickUpdateMenuItem>
```

When `supportsMultipleSelection` is `true`, it will extend the menu shown after pressing `Quick actions`, which means that menu items will show up in both the single selection and multiple selection menus. Otherwise, it will only extend the single update menu.

Closes #42737.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure the feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features` and the feature `product-variation-management` is enabled under `Features` tab from `/wp-admin/tools.php?page=woocommerce-admin-test-helper`.
3. Go to Products > Add New > Variations (`/wp-admin/admin.php?page=wc-admin&path=/add-product&tab=variations`) and add some attributes to generate variations.
4. Wait until variations get generated
5. Verify the actions under the vertical ellipsis are working well.
6. Select at least one variation and verify the Quick update actions are working well too.
7. Now it's time to test extensibility. You can checkout the branch `dev/42737_slotfill_for_variation_menus_test`. I added a few fields to different parts of the variation actions menus. [This extension](https://github.com/woocommerce/woocommerce/files/13973436/myplugin.zip) will add the following components


## Update

You can see the components [here](https://gist.github.com/octaedro/db58dd4377e1f57157080fd624925a28).

It adds:
- an item with the text `** Top level item 2 **` at the top level of the single update. It has order 20, so it should be rendered after the other top-level item.
- an item with the text `** Top level item **` at the top level of the single and multiple updates with order 10.
- an item with the text `** Pricing item / Set price **` into the pricing group for single and multiple updates. After pressing it, it changes the regular price for a random number. It works for single and multiple variations.
- an item with the text `** Shipping item **` into the shipping group of single updates. It doesn't trigger the onClose callback.
- an item with the text `** Downloads item / Set download limit **` into the downloads group of single updates. After pressing it, it updates the download item with a random number.
- an item with the text `** Inventory item / Set stock status: outofstock **` into the inventory group of single and multiple updates. After pressing it, it modifies the stock status to `out of stock`. It works for single and multiple variations.

Every added item will print in the browser's console:
```
[click] -> Plugin top level
Type - [ "single" or "multiple and single" ] items 
Selected items: [ selection ]
[ a quick description ]
```
<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

octaedro committed:
Add slotFills to main variation actions

octaedro committed:
Add slotFills to groups

octaedro committed:
Create variation-actions component

octaedro committed:
Add references

octaedro committed:
Remove file and change reference

octaedro committed:
Add Dropdown

octaedro committed:
Add styles

octaedro committed:
Add changelog

octaedro committed:
split variation-actions-menu

octaedro committed:
Add tests

octaedro committed:
Remove variation-actions-menu and variations-actions-menu

octaedro committed:
Rename variation-actions and add tests

octaedro committed:
Rename VariationSingleUpdateMenuItem

octaedro committed:
Rename single-update

octaedro commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-actions-menu-item.tsx:
```diff
@@ -0,0 +1,90 @@
+	);
+};
+
+export const VariationSingleUpdateMenuItem = ( {
```
I left the components `VariationSingleUpdateMenuItem` and `VariationQuickUpdateMenuItem` here because I considered them as one of the same kind as `VariationActionsMenuItem` (an extensibility component), but I'm totally ok with moving them to other files.
The idea underneath was to have multiple ways to do the same. The component `VariationActionsMenuItem` also provides the possibility of adding items under both action menus.

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-actions.tsx:
```diff
@@ -0,0 +1,150 @@
+					onClose={ onClose }
+				/>
+				<VariationActionsMenuItem.Slot
+					group={ '_main' }
```
I'm not sure we need/want this slot.

Why the `_` in the front of the group name?

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-actions.tsx:
```diff
@@ -0,0 +1,150 @@
+					/>
+				) }
+				<VariationActionsMenuItem.Slot
+					group={ '_secondary' }
```
I think we may not want to allow a slot at this level (or, at least not document/support it). The name of the group might be better as `top-level`.

We should check with @j111q to see where extensions should put menu items that don't fall under any of the existing groups, and whether they should be able to create their own groups.

octaedro replied:
Good call! I'll rename those groups initially to `top-level`, `secondary`, and `tertiary` and wait for Jill's answer to confirm if we should keep or remove those slots. I like the idea of providing a lot of flexibility, but as you said, maybe that would be too much.

mattsherman reviewed (changes requested):
Nice start on this, @octaedro ! 🙌

I'm interested in understanding why your implementation differs from the proposed API in #42737?

```jsx
<VariationQuickUpdateMenuItem
    group= { 'pricing' }
    supportsMultipleSelection={ true }
    order={ 20 }
>
    // Dropdown, MenuGroup, MenuItem
</VariationQuickUpdateMenuItem>
```

There are a few things I prefer about the original proposed API:

* There is a single fill, `<VariationQuickUpdateMenuItem>` that is used.
    * Having multiple ways of doing the same thing complicates the API, I believe.
    * I'd get rid of `VariationSingleUpdateMenuItem` and `VariationQuickUpdateMenuItem`)
* I think the name `<VariationQuickUpdateMenuItem>` is more specific for what this is used for.
    * The name `<VariationActionsMenuItem>` is a bit too vague in my opinion.
* I think having a boolean prop to set whether multiple selections are supported is easier to understand and harder to mess up (with a string prop, you might mistype it).
    * I do realize that this means that every menu item in the "multiple" menu appears in the single menu. I think that is reasonable and guarantees a level of consistency that should exist.

**Some other important points:**

* `VariationQuickUpdateMenuItem` needs to be exported from `@woocommerce/product-editor` so that it can be used by extensions!
* I think test instructions for extensibility should alway be done with an extension (if even just some code you put in a `mu-plugin` file) as opposed to modifying code (I personally will never test a PR by modifying code if at all possible, since that is not a real-world scenario and can introduce mistakes or hidden other issues, such as the above)

octaedro committed:
Add supportsMultipleSelection and slotFill rename

octaedro committed:
Add component export

octaedro committed:
Rename groups

octaedro commented:
Hi @mattsherman, thank you for your review.

> I think the name <VariationQuickUpdateMenuItem> is more specific for what this is used for.

Since we have a `Quick update` button, I feel it could be a little confusing to refer to both action types in the same way, mainly because they have small differences (like the preview possibility and the shortcuts to update stock or set a list price).
But I'm ok sticking with the original naming.

> The name <VariationActionsMenuItem> is a bit too vague in my opinion.

When I added the component `VariationActionsMenuItem`, I was thinking about something more generic to let 3PDs add elements, for example, in the `pricing` group for both action menus at the same time (not only one at a time). This is why I used `type` instead of `supportsMultipleSelection`. I think that in terms of usability, the original is more intuitive but has a shorter functionality. I'm good with prioritizing usability over functionality at this moment 👍

> I do realize that this means that every menu item in the "multiple" menu appears in the single menu. I think that is reasonable and guarantees a level of consistency that should exist.

Yes, the differences between them appear in the first group of items: the preview button and the shortcuts to update stock and set a list price.

> I think test instructions for extensibility should alway be done with an extension

Good idea! I created an extension to test the slotFills and added [a link to it](https://github.com/woocommerce/woocommerce/files/13916996/myplugin.zip) in the testing instructions.

octaedro commented:
@mattsherman I addressed the suggestions you made. Could you take another look at this PR?

mattsherman reviewed (changes requested):
Thanks for making the updates, @octaedro ! I was able to test with the following plugin code:

```js
wp.plugins.registerPlugin( 'example-variation-update-menu', {
    scope: 'woocommerce-product-block-editor',
    render: function() {
        return wp.element.createElement(
            wp.element.Fragment,
            {},
           
            // top level multiple selection
            wp.element.createElement(
                wc.productEditor.__experimentalVariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: true,
                },
                wp.element.createElement(
                    'p',
                    {},
                    'Plugin top level multiple selection!',
                )
            ),

            // secondary multiple selection
            wp.element.createElement(
                wc.productEditor.__experimentalVariationQuickUpdateMenuItem,
                {
                    group: 'secondary',
                    supportsMultipleSelection: true,
                },
                wp.element.createElement(
                    'p',
                    {},
                    'Plugin secondary multiple selection!',
                )
            ),

            // inventory multiple selection
            wp.element.createElement(
                wc.productEditor.__experimentalVariationQuickUpdateMenuItem,
                {
                    group: 'inventory',
                    supportsMultipleSelection: true,
                },
                wp.element.createElement(
                    'p',
                    {},
                    'Plugin inventory multiple selection!',
                )
            ),

            // top level single selection
            wp.element.createElement(
                wc.productEditor.__experimentalVariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: false,
                },
                wp.element.createElement(
                    'p',
                    {},
                    'Plugin top level single selection!',
                )
            ),

            // secondary single selection
            wp.element.createElement(
                wc.productEditor.__experimentalVariationQuickUpdateMenuItem,
                {
                    group: 'secondary',
                    supportsMultipleSelection: false,
                },
                wp.element.createElement(
                    'p',
                    {},
                    'Plugin secondary single selection!',
                )
            ),

            // inventory single selection
            wp.element.createElement(
                wc.productEditor.__experimentalVariationQuickUpdateMenuItem,
                {
                    group: 'inventory',
                    supportsMultipleSelection: false,
                },
                wp.element.createElement(
                    'p',
                    {},
                    'Plugin inventory single selection!',
                )
            ),
        );
    }
} );
```

It looks like you misunderstood how I envisioned `supportsMultipleSelection` working... I think it should work like the following:

* `supportsMultipleSelection === true`: Menu item shows up in *both* the single selection and multiple selection menus
* `supportsMultipleSelection === false`: Menu item shows up only in the single selection menu

Note: This doesn't offer a way to only have a menu item in multiple selection, but I can't think of a scenario where that would be needed, and I like that forced consistency. But, we could switch back to your approach if you think we should have the flexibility to place menu items only in the multiple selection menu.

Now, an improvement/change...

Testing this, and looking at how Gutenberg handles similar SlotFills, I think we should have an API that is a bit more opinionated in what can be filled.

We currently let the plugin fill in anything, as we both did in our examples (filling in a `p` element). This requires the plugin to know what type of element really should be used (`MenuItem`).

```js
<VariationQuickUpdateMenuItem group={ 'pricing } supportsMultipleSelection={ true }>
    <MenuItem onClick={handleClick}>
        Set sale price to % off
    </MenuItem>
</VariationQuickUpdateMenuItem>
```

Instead, I think we should do something similar to Gutenberg's [`PluginMoreMenuItem`](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-more-menu-item/), and wrap the fill automatically in a `MenuItem`...

```js
<VariationQuickUpdateMenuItem
    group={ 'pricing }
    supportsMultipleSelection={ true }
    onClick={handleClick}    
>
    Set sale price to % off
</VariationQuickUpdateMenuItem>
```

This will enforce more consistency in how menu items are rendered, and give us the flexibility to change how the menu is implemented in the future (for example, if we wanted to reimplement things to not use `MenuItem` under the hood).

mattsherman commented:
Also, I wonder if we should *not* mark this SlotFill as experimental, as we will be suggesting/documenting it for extensions to use.

octaedro committed:
Show multipleSelection items in both components

octaedro committed:
Take fills as MenuItems

octaedro committed:
Remove __experimental from slotFill

octaedro commented:
Thank you for reviewing the PR @mattsherman 

> This doesn't offer a way to only have a menu item in multiple selection, but I can't think of a scenario where that would be needed, and I like that forced consistency. But, we could switch back to your approach if you think we should have the flexibility to place menu items only in the multiple selection menu.

Yes, that makes sense to me. I added it in the commit [4f18d70](https://github.com/woocommerce/woocommerce/pull/43441/commits/4f18d70cd85673410b4cff46bd412ab7fef2618b)

> Instead, I think we should do something similar to Gutenberg's [PluginMoreMenuItem](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-more-menu-item/), and wrap the fill automatically in a MenuItem...

Good idea! I'm a little hesitant about removing flexibility here, but I think that it makes sense to simplify extensibility as much as possible. I added in the commit [e5383ff](https://github.com/woocommerce/woocommerce/pull/43441/commits/e5383ff190f75ff9a219343e55991dfdd02307d7)

octaedro committed:
Fix lint

octaedro commented:
Hi @mattsherman, I just added the changes you suggested. Could you take another look at this PR? Sorry for the back and forth.
I also updated [the testing plugin](https://github.com/woocommerce/woocommerce/files/13935089/myplugin.zip) to make it stick better with the latest changes.

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-quick-update-menu-item.tsx:
```diff
@@ -0,0 +1,98 @@
+
+const getMenuItem = ( children: React.ReactNode ) => (
+	<MenuItem>
+		<span>{ children }</span>
```
Is there a reason for the span? I think we should do without it if we can... let's keep the DOM as tidy as possible. 😄

mattsherman replied:
I'd suggest adding a unit test for `onClick` too.

octaedro replied:
> I'd suggest adding a unit test for onClick too.

Good idea! I added a few tests in the commit [dc7bccf](https://github.com/woocommerce/woocommerce/pull/43441/commits/dc7bccfc20be552a4dbd9bd308a11675f753ce2b)

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-quick-update-menu-item.tsx:
```diff
@@ -0,0 +1,98 @@
+};
+
+const getMenuItem = ( children: React.ReactNode ) => (
+	<MenuItem>
```
I tried using the following:

```js
wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: true,
                    onClick: () => alert( 'Plugin top level multiple selection!' ),
                },
                'Plugin top level multiple selection!',
),
```

And my `onClick` didn't work.

I assume it is because we don't actually pass down any props to the `MenuItem`.

mattsherman replied:
I'd say to keep things locked up a bit and only support that props we really need right now, like `onClick`. That way, we keep the flexibility to change the underlying implementation in the future. And, we prevent extensions from using props that would result in poor UX.

octaedro replied:
Added in [5903136](https://github.com/woocommerce/woocommerce/pull/43441/commits/59031365dfc7104f2b219ed12e8e93f1eacd4e3f)

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-quick-update-menu-item.tsx:
```diff
@@ -0,0 +1,98 @@
+ * Internal dependencies
+ */
+import { MenuItemProps } from './types';
+import { QUICK_UPDATE, SINGLE_UPDATE } from './constants';
```
I think these constants should be renamed to `MULTIPLE_UPDATE` and `SINGLE_UPDATE`.

octaedro replied:
I like it! [2a2b72a](https://github.com/woocommerce/woocommerce/pull/43441/commits/2a2b72ab663809c1e462229fe545d8cabddf46ab)

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/constants.ts:
```diff
@@ -0,0 +1,2 @@
+export const QUICK_UPDATE = 'quick-update';
```
I think this should be changed to `export const MULTIPLE_UPDATE = 'multiple-update';`

mattsherman reviewed (changes requested):
Things are coming together nicely on this! No need to apologize for the back-and-forth... that's just how it is sometimes!

Once the `onClick` handling is in, I'll give it another round of testing and review!

mattsherman commented:
> Things are coming together nicely on this! No need to apologize for the back-and-forth... that's just how it is sometimes!
> 
> Once the `onClick` handling is in, I'll give it another round of testing and review!

Oh yeah, just for reference, here's the latest test code I used:

```js
wp.plugins.registerPlugin( 'example-variation-update-menu', {
    scope: 'woocommerce-product-block-editor',
    render: function() {
        return wp.element.createElement(
            wp.element.Fragment,
            {},
           
            // top level multiple selection
            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: true,
                    order: 10,
                    onClick: () => alert( 'Plugin top level multiple selection!' ),
                },
                'Plugin top level multiple selection 1!',
            ),

            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: true,
                    order: 20,
                    onClick: () => alert( 'Plugin top level multiple selection!' ),
                },
                'Plugin top level multiple selection 2!',
            ),

            // secondary multiple selection
            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'secondary',
                    supportsMultipleSelection: true,
                },
                'Plugin secondary multiple selection!',
            ),

            // inventory multiple selection
            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'inventory',
                    supportsMultipleSelection: true,
                },
                'Plugin inventory multiple selection!',
            ),

            // top level single selection
            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: false,
                    order: 10,
                },
                'Plugin top level single selection 1!',
            ),

            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'top-level',
                    supportsMultipleSelection: false,
                    order: 15,
                },
                'Plugin top level single selection 2!',
            ),

            // secondary single selection
            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'secondary',
                    supportsMultipleSelection: false,
                },
                'Plugin secondary single selection!',
            ),

            // inventory single selection
            wp.element.createElement(
                wc.productEditor.VariationQuickUpdateMenuItem,
                {
                    group: 'inventory',
                    supportsMultipleSelection: false,
                },
                'Plugin inventory single selection!',
            ),
        );
    }
} );
```

octaedro committed:
Add onClick to slotFill

octaedro committed:
Rename constant QUICK_UPDATE, add new constant

octaedro committed:
Add tests for getGroupName and getMenuItem

octaedro committed:
Rename variation-actions tests

octaedro committed:
fix test description

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-actions.tsx:
```diff
@@ -0,0 +1,148 @@
+				<VariationQuickUpdateMenuItem.Slot
+					group={ 'secondary' }
+					supportsMultipleSelection={ supportsMultipleSelection }
+				/>
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-actions.tsx:
```diff
@@ -0,0 +1,148 @@
+				>
+					{ __( 'Delete', 'woocommerce' ) }
+				</MenuItem>
+				<VariationQuickUpdateMenuItem.Slot
+					group={ 'tertiary' }
+					supportsMultipleSelection={ supportsMultipleSelection }
+				/>
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-quick-update-menu-item.tsx:
```diff
@@ -0,0 +1,110 @@
+		: VARIATION_ACTIONS_SLOT_NAME;
+};
+
+export const getMenuItem = (
+	children: React.ReactNode,
+	onClick: () => void
+) => <MenuItem onClick={ onClick }>{ children }</MenuItem>;
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/quick-update-menu.tsx:
```diff
@@ -0,0 +1,50 @@
+import { VariationActionsMenuProps } from './types';
+import { VariationActions } from './variation-actions';
+
+export function QuickUpdateMenu( {
```
I think we should refer to this as the `MultipleUpdateMenu`.

octaedro replied:
Done in commit [71779a5](https://github.com/woocommerce/woocommerce/pull/43441/commits/71779a54331ed63185bc56f3c096ade1040f0363)

mattsherman commented on packages/js/product-editor/src/components/variations-table/pricing-menu-item/pricing-menu-item.tsx:
```diff
@@ -434,6 +436,12 @@ export function PricingMenuItem( {
 						>
 							{ __( 'Schedule sale', 'woocommerce' ) }
 						</MenuItem>
+						<VariationQuickUpdateMenuItem.Slot
+							group={ 'pricing' }
+							supportsMultipleSelection={
+								supportsMultipleSelection
+							}
+						/>
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/shipping-menu-item/shipping-menu-item.tsx:
```diff
@@ -239,6 +241,10 @@ export function ShippingMenuItem( {
 					>
 						{ __( 'Set weight', 'woocommerce' ) }
 					</MenuItem>
+					<VariationQuickUpdateMenuItem.Slot
+						group={ 'shipping' }
+						supportsMultipleSelection={ supportsMultipleSelection }
+					/>
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/inventory-menu-item/inventory-menu-item.tsx:
```diff
@@ -233,6 +235,12 @@ export function InventoryMenuItem( {
 						>
 							{ __( 'Edit low stock threshold', 'woocommerce' ) }
 						</MenuItem>
+						<VariationQuickUpdateMenuItem.Slot
+							group={ 'inventory' }
+							supportsMultipleSelection={
+								supportsMultipleSelection
+							}
+						/>
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/downloads-menu-item/downloads-menu-item.tsx:
```diff
@@ -199,6 +201,10 @@ export function DownloadsMenuItem( {
 					>
 						{ __( 'Set download expiry', 'woocommerce' ) }
 					</MenuItem>
+					<VariationQuickUpdateMenuItem.Slot
+						group={ 'downloads' }
+						supportsMultipleSelection={ supportsMultipleSelection }
+					/>
```
We need to pass through the `selection` and `onChange` so that fills can access that in their `onClick` handler and actually do something!

We should also pass  the`onClose` callbacks, as that will be needed too for extensions to close the menu when they have done their action!

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/test/variation-quick-update-menu-item.test.tsx:
```diff
@@ -0,0 +1,58 @@
+		expect( queryByText( 'My button' ) ).toBeInTheDocument();
+	} );
+
+	it( 'should call onClick when the menuItem is clicked', async () => {
+		const { getByRole } = render(
+			getMenuItem( <button>My button</button>, onClickMock )
+		);
+		await fireEvent.click( getByRole( 'button', { name: 'My button' } ) );
+		expect( onClickMock ).toHaveBeenCalled();
+	} );
```
It is confusing having a button in this test, as the `MenuItem` itself is a `<button>` with `role="menuitem"`. I think we should be using those when testing.

octaedro replied:
I modified the test in the commit [b54966b](https://github.com/woocommerce/woocommerce/pull/43441/commits/b54966b28828d255b3c984f9e6500a24c66f2223)

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/test/variation-quick-update-menu-item.test.tsx:
```diff
@@ -0,0 +1,58 @@
+/**
```
Love the tests! ❤️

I wonder if we could go a bit higher level and test:

- That using the fill actually fills the correct slot with a `MenuItem`
- The the callbacks are called (you are doing this for `onClick` already)

For example, currently, somebody could change `getMenuItem` so that it returns `null` and these tests would still pass, when clearly something is very wrong in that case.

Also, somebody could change the implementation to not have the `getMenuItem` function, and these unit tests would have to change, when the `getMenuItem` function is really just a small implementation detail that should be able to be changed/renamed without having to change unit tests.

octaedro replied:
> I wonder if we could go a bit higher level and test:
That using the fill actually fills the correct slot with a MenuItem
The the callbacks are called (you are doing this for onClick already)

Done in commit [b54966b](https://github.com/woocommerce/woocommerce/pull/43441/commits/b54966b28828d255b3c984f9e6500a24c66f2223)

mattsherman commented on packages/js/product-editor/src/components/variations-table/variation-actions-menus/variation-quick-update-menu-item.tsx:
```diff
@@ -0,0 +1,110 @@
+	supportsMultipleSelection,
+	onClick = () => {},
+} ) => {
+	if ( supportsMultipleSelection ) {
+		return (
+			<>
+				{ [ MULTIPLE_UPDATE, SINGLE_UPDATE ].map( ( updateType ) => (
+					<Fill
+						key={ updateType }
+						name={ getGroupName(
+							group,
+							updateType === MULTIPLE_UPDATE
+						) }
+					>
+						{ ( fillProps: Fill.Props ) =>
+							createOrderedChildren(
+								getMenuItem( children, onClick ),
+								order,
+								fillProps
+							)
+						}
+					</Fill>
+				) ) }
+			</>
+		);
+	}
+	return (
+		<Fill name={ getGroupName( group, supportsMultipleSelection ) }>
+			{ ( fillProps: Fill.Props ) =>
+				createOrderedChildren(
+					getMenuItem( children, onClick ),
+					order,
+					fillProps
+				)
+			}
+		</Fill>
+	);
```
It would be nice to rework this to avoid the code duplication and multiple returns. It's a little confusing to look at, and having the `<Fill>` twice in this function could lead to bugs down the road if a change was made in one path and not the other (say if another prop needed to be passed to the children of the fill.

octaedro replied:
Done in [c1d9901](https://github.com/woocommerce/woocommerce/pull/43441/commits/c1d9901ae213acb5071611a5ad2e198a4b4ff30a)

mattsherman reviewed (changes requested):
@octaedro Sorry for not catching the obvious omission earlier of not providing the fills with the necessary information in order to actually do something! It wasn't clear when there were other things in the way/missing.

I think once that is in place, we will be good to go.

One thing I'd suggest is providing some sample/test code that is functional, to help verify that the SlotFill will meet needs of extensions. This could be be having the extension's menu item console.log the selection and/or actually call `onChange` to do something obvious to the selected variations.

octaedro committed:
Add onChange, onClose and selection to slots

octaedro committed:
Refactor slotFill

octaedro committed:
Refactor slotFill file

octaedro committed:
Rename QuickUpdateMenu to MultipleUpdateMenu

octaedro committed:
Modify tests

octaedro committed:
Fix test name

octaedro committed:
Improve import

octaedro commented:
Thank you @mattsherman for the review. I addressed the changes you suggested and added more tests.

> One thing I'd suggest is providing some sample/test code that is functional, to help verify that the SlotFill will meet needs of extensions. This could be be having the extension's menu item console.log the selection and/or actually call onChange to do something obvious to the selected variations.

Sure thing, I updated [this testing plugin](https://github.com/woocommerce/woocommerce/files/13973436/myplugin.zip) to get that. Just download it and install it on your end and it should work.

Could you take another look at this PR?

mattsherman reviewed (changes requested):
@octaedro Thanks for all the updates on this one. I think there has been a large amount of back-and-forth because we honestly didn't have the issue ready to start when we started it! In retrospect, we should have made sure we fully fleshed out the following before we began:

- UX, including every little detail
- API; by including full examples of the extension code we wanted to write in the issue, we would have made sure we handled many important points that have only come out over multiple rounds of review/testing

And, we could have separated out the refactoring of the menus to avoid duplication in a separate PR before adding on the slot fills. That could have preventing the PR from being open for so long and growing so much. But, we often don't see that at the beginning!

Okay, on to the latest round of testing feedback:

Overall, things are getting very close!

Dev experience:

- It feels cumbersome that `selection` can either be an object or an array. It complicates the code that an extension needs to write for the (anticipated common case) of multiple selection support. I think it would be nicer from an extension standpoint if `selection` was always an array, even when accessing the menu item through the single menu.
- Relatedly, it feels cumbersome that `onChange` takes either an object or an array, depending on what `selection` is. It would be simpler if `onChange` always took an array.

UX:

In the Pricing menu, we have two groups, `List Price` and `Sale Price`. This makes it look confusing when an extension adds a menu item to this menu, since it looks like it is part of the `Sale Price` group. Perhaps we should always add a separator before any extension menu items? For consistency, let's do that for any menu item. Including the top level slot and the secondary slot.

<img width="1089" alt="Screenshot 2024-01-18 at 10 28 20" src="https://github.com/woocommerce/woocommerce/assets/2098816/fb51b8f5-661d-4079-b3cc-462bad844aef">

Side note:

Your need to set the `sku` in your test code is actually because of a bug in our quick actions handling, which I found and documented today...

- https://github.com/woocommerce/woocommerce/issues/43800

octaedro committed:
Always return array of selected items

octaedro committed:
Create MenuGroup for every extension menu items

octaedro committed:
Group fills

mattsherman reviewed (approved):
Re-tested and appears to work as expected (I'll admit, I'm getting some test/review fatigue on this! 😄).

Thanks for making all the changes. I think this looks like it is in a good state to merge.

Note that I can't actually make any quick action updates, but that is because of https://github.com/woocommerce/woocommerce/issues/43800

> I think that is not a problem whether the `onChange` accepts single items and arrays. I would continue analyzing that possibility because the single-item update is already implemented and makes the function more flexible. If it's ok with you, I would tackle that in a follow-up PR.

Okay.

> Aiming to wrap up this PR, I created a follow-up PR (it's a draft so far) with a few changes to allow a better integration with the variation actions menu. I'm thinking about the chance for 3PDs to add a whole group (a `dropdown` component as the other groups) to the menu. If we only allow them to add MenuItems, that will force them to add some hacky CSS to have a better experience; that happens because the new group will be rendered inside a `MenuItem`.

I'm not sure we want to allow extensions to add entire groups, but I think that we probably should, because even if we discourage it, it would be better for them to have a supported way to properly add a group instead of hacking a solution. Sounds good.

leewillis77 commented:
@octaedro The sample plugin you linked to only contains the built JS, not the original source which makes it difficult to see an example of how to use this API. Please can you post a copy that includes the JS source?

octaedro commented:
Hi @leewillis77!

> Please can you post a copy that includes the JS source?

Sure thing, no problem. I created a project with my plugin's code [here](https://github.com/octaedro/testing-plugin). Just clone the repo **in the WooCommerce root folder** because [we will need to reference](https://github.com/octaedro/testing-plugin/blob/main/package.json#L25) the product-editor with the changes in this branch. Then run `npm install` (you'll probably need to add `--legacy-peer-deps` at the end) and then `npm run build`.
After that symlink the folder to `wp-content/plugins` to see (and activate) the plugin.

Please let me know if the plugin doesn't work well for you.

leewillis77 commented:
@octaedro @mattsherman 

I haven't had chance to attempt porting this into our extensions, but from a quick inspection of the code presented here it's already making me quite uncomfortable. Our total code for registering 2 quick actions, *and* handling their callbacks currently is ~50 lines,

e.g. [paraphrased code, so please exclude typos etc]. 

```js
jQuery(function() {
	jQuery(document).on('woocommerce_variations_loaded', function() {
		if (jQuery( 'select.variation_actions' ).hasClass('woocommerce-msrp-ajax-attached')) {
			return;
		}
		jQuery( 'select.variation_actions' ).on( 'msrp_set_prices_ajax_data', function( e, data ) {
			value = window.prompt( woocommerce_admin_meta_boxes_variations.i18n_enter_a_value );
			if ( value != '' && value != null ) {
				data.value = value;
			}
			return data;
		});
		jQuery( 'select.variation_actions' ).addClass('woocommerce-msrp-ajax-attached');
	});
});
```

```php

add_action( 'woocommerce_variable_product_bulk_edit_actions', function() {
	?>
	<optgroup label="<?php esc_html_e( 'MSRP Prices', 'woocommerce-msrp' ); ?>">
		<option value="msrp_set_prices"><?php esc_html_e( 'Set prices', 'woocommerce-msrp' ); ?></option>
		<option value="msrp_clear_prices"><?php esc_html_e( 'Clear MSRP prices', 'woocommerce-msrp' ); ?></option>
	</optgroup>
	<?php
} );

add_action( 'woocommerce_bulk_edit_variations', function ( $bulk_action, $data, $product_id, $variations ) {
		switch ( $bulk_action ) {
			case 'msrp_set_prices':
				return bulk_action_set_prices( $data, $product_id, $variations );
				break;
			case 'msrp_clear_prices':
				return bulk_action_clear_prices( $data, $product_id, $variations );
				break;
		}
	},
	10,
	4
);

function bulk_action_set_prices( $data, $product_id, $variations ) {
	if ( ! isset( $data['value'] ) ) {
		return;
	}
	foreach ( $variations as $variation_id ) {
		$variation = wc_get_product( $variation_id );
		$variation->update_meta_data( '_msrp', $data['value'] );
		$variation->save();
	}
}

function bulk_action_clear_prices( $data, $product_id, $variations ) {
	foreach ( $variations as $variation_id ) {
		$variation = wc_get_product( $variation_id );
		$variation->delete_meta_data( '_msrp' );
		$variation->save();
	}
}
```

The solution being proposed here seems *significantly* more complex - probably a couple of 100 lines of code just to get the options displaying never mind handling their callbacks (related: is there any infrastructure for that?). Am I missing the point?

leewillis77 commented:
OK. So I've had a further look at porting our variation quick actions across to this.

For some context, here's how our actions are presented in the existing bulk actions menu:
![Screenshot 2024-01-30 at 11 57 23](https://github.com/woocommerce/woocommerce/assets/1097338/c64d8f9c-8ae4-4bed-8163-3839d1c29446)

As you can see the actions relating to our extensions data (MSRP pricing in this case) are grouped together for ease of use. 

Here's the code we're currently working on in an attempt to replicate this in the new product editor:
```js
import { registerPlugin } from '@wordpress/plugins';
import { Fragment } from '@wordpress/element';
import { MenuGroup, MenuItem } from '@wordpress/components';
import { VariationQuickUpdateMenuItem } from '@woocommerce/product-editor';
import { __ } from '@wordpress/i18n';

registerPlugin( 'example-variation-update-menu', {
	scope: 'woocommerce-product-block-editor',
	render: function () {
		return (
			<Fragment>
					<VariationQuickUpdateMenuItem group="pricing" supportsMultipleSelection={ true }>
						<MenuGroup label={ __( 'MSRP', 'woocommerce_msrp' ) }>
							<MenuItem>
								{ __( 'Set MSRP prices', 'woocommerce_msrp' ) }
							</MenuItem>
							<MenuItem>
								{ __( 'Clear MSRP prices', 'woocommerce_msrp' ) }
							</MenuItem>
						</MenuGroup>
					</VariationQuickUpdateMenuItem>
			</Fragment>
		);
	}
} );
```

Unfortunately this creates a somewhat broken layout (I think maybe this was what @mattsherman was referring to in https://github.com/woocommerce/woocommerce/pull/43441#pullrequestreview-1837427350 ?):
![Screenshot 2024-01-30 at 12 02 39](https://github.com/woocommerce/woocommerce/assets/1097338/b695d65f-8876-475e-acf9-0983cae28be9)

If we want to create a nice UX, then I think it's critical that extensions can group their actions for clarity/convenience. 

There are some other minor issues here, relating to single vs multiple selections (which isn't an issue on the current incarnation since the actions only apply to multiple items), in that ideally we'd call the action "Set price" when the menu is accessed against a single variant, but "Set prices" when accessed for multiple variations, but it doesn't seem that that is possible?

mattsherman commented:
> If we want to create a nice UX, then I think it's critical that extensions can group their actions for clarity/convenience.

We will be adding support for this in the future.

> There are some other minor issues here, relating to single vs multiple selections (which isn't an issue on the current incarnation since the actions only apply to multiple items), in that ideally we'd call the action "Set price" when the menu is accessed against a single variant, but "Set prices" when accessed for multiple variations, but it doesn't seem that that is possible?

This is currently not possible. And, as you can see from the built-in actions, we do not change the menu text based on the selection. But, this is something we should consider for a future improvement!

leewillis77 commented:
> We will be adding support for this in the future.

Thanks - is there an issue I can follow for that?

mattsherman commented:
> > We will be adding support for this in the future.
> 
> Thanks - is there an issue I can follow for that?

@leewillis77 Not yet, but I will create an issue in the next day and will link to it from here

leewillis77 commented:
Hi @mattsherman;

Any update on the requirement to add groups of quick actions in the new product editor as per your last comment?

I'm also seeing problems getting these slotfills working when building woocommerce myself. If I download and run WooCommerce 8.6.1 from wordpress.org, then my variation quick actions show up in the menu. If I install woocommerce from git, then my actions are missing from the drop-down (no obvious errors that would point to the cause). 

Here's the steps I'm using to install woocommerce from git - maybe I'm missing something?

- `git clone  git@github.com:woocommerce/woocommerce.git woocommerce-repo`
- `cd woocommerce-repo`
- `nvm use`
- `pnpm install`
- `pnpm build`
- `cd ..`
- `ln -s woocommerce-repo/plugins/woocommerce`
- `wp plugin activate woocommerce`

mattsherman commented:
> Any update on the requirement to add groups of quick actions in the new product editor as per your last comment?

Unfortunately, no update yet. I will get that issue up soon though, so we can track it better.

> I'm also seeing problems getting these slotfills working when building woocommerce myself. If I download and run WooCommerce 8.6.1 from wordpress.org, then my variation quick actions show up in the menu. If I install woocommerce from git, then my actions are missing from the drop-down (no obvious errors that would point to the cause).

> Here's the steps I'm using to install woocommerce from git - maybe I'm missing something?

Hmm, those steps look good to me. One note, these fills were mistakenly added with the `__experimental` prefix initially, and then corrected to have the `__experimental` prefix. It looks like unfortunately, the non-prefixed naming made it into WooCommerce 8.6.

So, try using `__experimentalVariationQuickUpdateMenuItem`. Let us know any errors that appear in the dev console, in case they might help us troubleshoot things.

leewillis77 commented:
> So, try using __experimentalVariationQuickUpdateMenuItem. Let us know any errors that appear in the dev console, in case they might help us troubleshoot things.

No change (except, as expected it then doesn't work in WC 8.6.1 either). The console logs are attached below, but as far as I can tell these are largely the same on the release, or locally-built version, and the same whether or not my code is active. 

[wc.test-1709913542342.log](https://github.com/woocommerce/woocommerce/files/14540328/wc.test-1709913542342.log)

leewillis77 commented:
For info I still can't get this working with 8.7 (either built from git, or downloaded from wordpress.org). @octaedro's test plugin doesn't seem to work either as far as I can tell.

leewillis77 commented:
Just following up. After an afternoon of wrangling I've now got quick update actions showing up again. I've made a bunch of changes from my (previously working) code, and not sure if all of those were necessary or not, however I can now add quick update actions (using __experimentalVariationQuickUpdateMenuItem).

I think we'd still like groupings before we could ship this live otherwise it just looks confusing.

louwie17 commented:
Hey @leewillis77, sorry for the delay on this, the follow up on this fell through the cracks.
We created an issue for this, which you can follow: https://github.com/woocommerce/woocommerce/issues/47092
Also if you feel comfortable, we are more than happy to review it as well if you wanted to implement it :) But if not it is in our queue.
